### PR TITLE
Add remark-language-server

### PR DESCRIPTION
**Pull request by remcohaszing**
This is a language server for linting and formatting markdown files using remark.

I used `typescript-language-server` and `yaml-language-server` for reference. Any feedback is welcome. :)

/cc @wooorm @ChristianMurphy @Murderlon

### CODE REVIEW

**Review by ChristianMurphy — APPROVED** [review]

**Comment by mattn** [comment]
I'm not good at this but What is different from unified-language-server ?

**Comment by remcohaszing** [comment]
- `unified` is a toolchain which supports processing files using ASTs.
- `unified-engine` is a wrapper around `unified` which supports a standardized way to handle configuration files and ignore files which allows to create integrations, for example the various CLIs (`remark-cli`, `rehype-cli`) in the unified ecosystem are based on it.

I recently changed `unified-language-server` from a language server which implements custom logic for loading configurations to a library for creating language servers based on `unified-engine`. `remark-language-server` is the first implementation of this.

This means `remark-language-server` behaves the same as `remark-cli`, but as as a language server (provides linting and formatting capabilities based on `.remarkrc` and `.remarkignore`).

Language servers for more unified based ecosystems are coming soon.

---

TL;DR `unified-language-server` is a library for creating unified based language servers. `remark-language-server` is an implementation of `unified-language-server` which supports markdown.

**Comment by mattn** [comment]
Thanks your explaining. Sorry, how to use this? Probably I'm having wrong configuration?

image

File: settings.json [hunk 1]
```diff
@@ -644,6 +644,14 @@
       "requires": []
     }
   ],
+  "markdown": [
+    {
+      "command": "remark-language-server",
+      "requires": [
+        "npm"
+      ]
+    }
```
Comment (mattn) [inline settings.json:653]:
```
"requires": [
  "npm",
  "typescript-language-server",
  "yaml-language-server"
]
```

Comment (remcohaszing) [reply settings.json:653]: > I used `typescript-language-server` and `yaml-language-server` for reference.

With this I meant I looked how those are implemented to see how I should implement `remark-language-server`. Apart from that they’re not related / required.

**Comment by remcohaszing** [comment]
This project uses a [remark configuration file](https://github.com/remarkjs/remark/tree/main/packages/remark-cli#example-config-files-json-yaml-js).

For example, you could clone and setup https://github.com/unifiedjs/unified. This project has the `"remarkConfig"` property in `package.json`.

```sh
git clone git@github.com:unifiedjs/unified.git
cd unified
npm install
```

Now create a file named `foo.md` with the following contents:

```markdown
This line is toooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo long.

This project is based on `remark-preset-wooorm`. Also there should be _two_ spaces between sentences.
```

Now I see the following

image

This is based on the following snippet from `~/.vim/vimrc`:

```viml
if (executable('remark-language-server'))
  au User lsp_setup call lsp#register_server({
  \ 'name': 'remark',
  \ 'cmd': {server_info->['remark-language-server', '--stdio']},
  \ 'allowlist': ['markdown'],
  \ })
endif
```

**Comment by mattn** [comment]
My installation instruction 
```
git clone git@github.com:unifiedjs/unified.git
cd unified
npm install
```
But 
image

File: settings/remark-language-server.vim [hunk 2]
```diff
@@ -0,0 +1,14 @@
+augroup vim_lsp_settings_remark_language_server
+  au!
+  LspRegisterServer {
+      \ 'name': 'remark-language-server',
+      \ 'cmd': {server_info->lsp_settings#get('remark-language-server', 'cmd', [lsp_settings#exec_path('typescript-language-server')]+lsp_settings#get('remark-language-server', 'args', ['--stdio']))},
```
Comment (mattn) [inline settings/remark-language-server.vim:5]: This `typescript-language-server` is correctly?

Comment (mattn) [reply settings/remark-language-server.vim:5]: Now I got this working. Could you please fix `s/typescript/remark/` ?
image

Comment (mattn) [inline settings/remark-language-server.vim:5]:
```suggestion
      \ 'cmd': {server_info->lsp_settings#get('remark-language-server', 'cmd', [lsp_settings#exec_path('remark-language-server')]+lsp_settings#get('remark-language-server', 'args', ['--stdio']))},
```

**Comment by mattn** [comment]
Thank you

File: settings/remark-language-server.vim (hunk 2 above)
Comment (remcohaszing) [reply settings/remark-language-server.vim]: Awesome! Thanks for helping troubleshoot
